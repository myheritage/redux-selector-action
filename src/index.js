const REDUX_ACTION_SELECTOR = '@@redux_action_selector';
const PLACEHOLDER = `${REDUX_ACTION_SELECTOR}/placeholder`;

/**
 * A selector function which returns the "placeholder" constant
 *
 * @returns {string} "placeholder" value
 */
export const getPlaceholder = () => PLACEHOLDER;

/**
 * A middleware which handles any action from type "redux-action-selector" in order to inject the store selectors output,
 * to the given action creator inside the handled action payload.
 * Afterwards, the middleware dispatch the new action with its injected args without passing them, but to calculate them against the store.
 * Otherwise calls next middleware and return its output.
 *
 * @param {object} store - The redux store API
 * @returns {function} - A callback which is called with the next middleware
 */
export function reduxActionSelectorMiddleware(store) {
    return next => action => {
        const {type, payload, meta} = action;
		
        if (type === REDUX_ACTION_SELECTOR) {
            throwIfNotFunction(payload);
            const {selectors, args} = meta;
            const state = store.getState();
            const computedSelectors = selectors.map(selector => isFunction(selector) ? selector(state) : selector);
            const actionCreator = partial(payload, ...computedSelectors);
            store.dispatch(actionCreator(...args));
        } else {
            return next(action);
        }
    }
}

/**
 * Accepts an action creator and selectors which we want to inject as dependencies of the action creator.
 * A dependency can be a selector function, which its output will be passed or any other value.
 * Supports the following formats:
 * 1. createActionSelector(dep1, dep2, ..., actionCreator)
 * 2. createActionSelector([dep1, dep2, ...], actionCreator)
 * Returns a new action creator which we can dispatch with redux store.dispatch.
 * The new action creator can be tested using its new props resultFunc, dependencies.
 * resultFunc = the given action creator function
 * dependencies = the given selectors as an array
 *
 * We can use a special selector as a dependency, called getPlaceholder, for supporting args of the returned action creator.
 * Each getPlaceholder saves the argument position for another outer argument.
 * Once we dispatch the new action, its args will fill all the placeholders by their position order.
 *
 * @param  {...function} funcs - selectors and action creator
 * @returns {function} action selector
 */
export function createActionSelector(...funcs) {
    const actionCreator = funcs.pop();
    const selectors = Array.isArray(funcs[0]) ? funcs[0] : funcs;
    throwIfNotFunction(actionCreator);

    const actionSelector = (...args) => ({
        type: REDUX_ACTION_SELECTOR,
        payload: actionCreator,
        meta: {
            selectors,
            args,
        },
    });

    actionSelector.resultFunc = actionCreator;
    actionSelector.dependencies = selectors;
    return actionSelector;
}

/**
 * Returns new function which is partially applied.
 * Meaning part of the args will be bound to the function and the function won't accept them anymore, in order to reduce its arguments.
 * In case we get a bound arg which is the "placeholder", we inject to the new function one of the args which where applied,
 * instead of the placeholder, by their position.
 *
 * @param {function} func - Function to bind its args
 * @param {array} boundArgs - The arguments we want to bind to the given function
 * @returns {function} - partial function
 */
function partial(func, ...boundArgs) {
    return (...args) => {
        let argIndex = 0;
        const newArgs = boundArgs
            .map(arg => arg === PLACEHOLDER ? args[argIndex++] : arg)
            .concat(args.slice(argIndex));
        return func(...newArgs);
    }
}

/**
 * Returns whether the given arg is a function.
 *
 * @param {*} func - A potential function
 * @returns {boolean} - Whether is function
 */
function isFunction(func) {
    return typeof func === 'function';
}

/**
 * Throws an error if the given arg is not a function, otherwise continue.
 *
 * @param {function} func - The function to verify
 * @returns {undefined}
 * @throws TypeError
 */
function throwIfNotFunction(func) {
    if (!isFunction(func)) {
        throw new TypeError('Action selectors should be based on action creator of type function, ' +
            `instead received the following type: [${typeof func}]`);
    }
}