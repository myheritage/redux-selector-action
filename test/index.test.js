import {createSelectorAction, getPlaceholder, reduxSelectorActionMiddleware} from '../src';

describe('redux selector action', () => {
    describe('getPlaceholder', () => {
        it('should return placeholder object', () => {
            expect(typeof getPlaceholder()).toBe('object');
        });

        it('should match placeholder object empty', () => {
            expect(getPlaceholder()).toMatchSnapshot();
        });

        it('should throw error for invalid selectorObj', () => {
            expect(() => getPlaceholder('wrong type')).toThrowErrorMatchingSnapshot();
        });

        it('should throw error for nullable selectorObj', () => {
            expect(() => getPlaceholder(null)).toThrowErrorMatchingSnapshot();
        });

        it('should match placeholder object with selector object', () => {
            expect(getPlaceholder({dep: () => 'myDep'})).toMatchSnapshot();
        });
    });

    describe('createSelectorAction', () => {
        it('should throw error for invalid resultFunc', () => {
            expect(createSelectorAction).toThrowErrorMatchingSnapshot();
        });

        it('should return selector actions without dependencies', () => {
            const selectorAction = createSelectorAction(jest.fn());
            expect(selectorAction.resultFunc).toBeDefined();
            expect(selectorAction.dependencies).toHaveLength(0);
            expect(selectorAction('myArg')).toMatchSnapshot();
        });

        it('should return selector action with dependencies as args', () => {
            const selectorAction = createSelectorAction(() => 'myDep1', () => 'myDep2', jest.fn());
            expect(selectorAction.resultFunc).toBeDefined();
            expect(selectorAction.dependencies).toHaveLength(2);
            expect(selectorAction('myArg')).toMatchSnapshot();
        });

        it('should return selector action with array of dependencies', () => {
            const selectorAction = createSelectorAction([
                () => 'myDep1',
                () => 'myDep2',
            ], jest.fn());
            expect(selectorAction.resultFunc).toBeDefined();
            expect(selectorAction.dependencies).toHaveLength(2);
            expect(selectorAction('myArg')).toMatchSnapshot();
        });
    });

    describe('reduxSelectorActionMiddleware', () => {
        let actionCreator, dispatch, getState, handleAction, next, selectorAction;

        beforeEach(() => {
            dispatch = jest.fn();
            getState = jest.fn();
            next = jest.fn();
            handleAction = reduxSelectorActionMiddleware({getState, dispatch})(next);
            actionCreator = jest.fn((...args) => ({
                type: 'dummy_action',
                args,
            }));
            selectorAction = {
                type: '@@redux_selector_action',
                payload: actionCreator,
                meta: {
                    selectors: [],
                    args: [],
                },
            };
        });

        it('should call next for unknown action', () => {
            selectorAction.type = 'unknown';
            const nextReturnValue = true;
            next.mockReturnValue(nextReturnValue);
            const result = handleAction(selectorAction);
            expect(next).toHaveBeenCalledWith(selectorAction);
            expect(result).toBe(nextReturnValue);
        });

        it('should throw error for invalid payload type', () => {
            selectorAction.payload = undefined;
            expect(() => handleAction(selectorAction)).toThrowErrorMatchingSnapshot();
            expect(actionCreator).not.toHaveBeenCalled();
            expect(dispatch).not.toHaveBeenCalled();
        });

        it('should dispatch selector action without selectors and args', () => {
            handleAction(selectorAction);
            expect(actionCreator).toHaveBeenCalled();
            expect(dispatch).toHaveBeenCalled();
            expect(actionCreator.mock.calls[0]).toHaveLength(0);
        });

        it('should inject selectors output and non-function values to the given action creator', () => {
            selectorAction.meta.selectors.push(
                () => 'dep1',
                'constant',
                getPlaceholder,
            );
            handleAction(selectorAction);
            expect(dispatch).toHaveBeenCalled();
            expect(actionCreator.mock.calls).toMatchSnapshot();
        });

        it('should inject args to placeholders by position order', () => {
            selectorAction.meta.selectors.push(
                () => 'dep1',
                getPlaceholder,
                'const1',
                () => 'dep2',
                getPlaceholder(),
                () => 'dep3',
                'const2',
            );
            selectorAction.meta.args.push('arg1', 'arg2', 'arg3', 'arg4');
            handleAction(selectorAction);
            expect(dispatch).toHaveBeenCalled();
            expect(actionCreator.mock.calls).toMatchSnapshot();
        });

        it('should inject object to placeholders by position order', () => {
            selectorAction.meta.selectors.push(
                () => 'dep1',
                getPlaceholder({prop1: () => 'prop1Value'}),
                'const1',
                () => 'dep2',
                getPlaceholder({otherProp1: () => 'otherProp1Value'}),
                () => 'dep3',
                'const2',
            );
            selectorAction.meta.args.push(
                {prop3: 'myProp3Value', prop2: 'myProp2Value'},
                {otherProp3: 'myOtherProp3Value', otherProp2: 'myOtherProp2Value'},
                'arg3',
                'arg4',
            );
            handleAction(selectorAction);
            expect(dispatch).toHaveBeenCalled();
            expect(actionCreator.mock.calls).toMatchSnapshot();
        });
    });
});