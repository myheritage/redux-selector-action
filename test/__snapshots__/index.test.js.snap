// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux selector action createSelectorAction should return selector action with array of dependencies 1`] = `
Object {
  "meta": Object {
    "args": Array [
      "myArg",
    ],
    "selectors": Array [
      [Function],
      [Function],
    ],
  },
  "payload": [MockFunction],
  "type": "@@redux_selector_action",
}
`;

exports[`redux selector action createSelectorAction should return selector action with dependencies as args 1`] = `
Object {
  "meta": Object {
    "args": Array [
      "myArg",
    ],
    "selectors": Array [
      [Function],
      [Function],
    ],
  },
  "payload": [MockFunction],
  "type": "@@redux_selector_action",
}
`;

exports[`redux selector action createSelectorAction should return selector actions without dependencies 1`] = `
Object {
  "meta": Object {
    "args": Array [
      "myArg",
    ],
    "selectors": Array [],
  },
  "payload": [MockFunction],
  "type": "@@redux_selector_action",
}
`;

exports[`redux selector action createSelectorAction should throw error for invalid resultFunc 1`] = `"Selector actions should be based on action creator of type function, instead received the following type: [undefined]"`;

exports[`redux selector action getPlaceholder should match placeholder object empty 1`] = `
Object {
  "placeholder": "@@redux_selector_action/placeholder",
  "selectorObj": undefined,
}
`;

exports[`redux selector action getPlaceholder should match placeholder object with selector object 1`] = `
Object {
  "placeholder": "@@redux_selector_action/placeholder",
  "selectorObj": Object {
    "dep": [Function],
  },
}
`;

exports[`redux selector action getPlaceholder should throw error for invalid selectorObj 1`] = `"Placeholder for object arg should be of type object and not null, instead received the following type: [string]"`;

exports[`redux selector action getPlaceholder should throw error for nullable selectorObj 1`] = `"Placeholder for object arg should be of type object and not null, instead received the following type: [object]"`;

exports[`redux selector action reduxSelectorActionMiddleware should inject args to placeholders by position order 1`] = `
Array [
  Array [
    "dep1",
    "arg1",
    "const1",
    "dep2",
    "arg2",
    "dep3",
    "const2",
    "arg3",
    "arg4",
  ],
]
`;

exports[`redux selector action reduxSelectorActionMiddleware should inject object to placeholders by position order 1`] = `
Array [
  Array [
    "dep1",
    Object {
      "prop1": "prop1Value",
      "prop2": "myProp2Value",
      "prop3": "myProp3Value",
    },
    "const1",
    "dep2",
    Object {
      "otherProp1": "otherProp1Value",
      "otherProp2": "myOtherProp2Value",
      "otherProp3": "myOtherProp3Value",
    },
    "dep3",
    "const2",
    "arg3",
    "arg4",
  ],
]
`;

exports[`redux selector action reduxSelectorActionMiddleware should inject selectors output and non-function values to the given action creator 1`] = `
Array [
  Array [
    "dep1",
    "constant",
    undefined,
  ],
]
`;

exports[`redux selector action reduxSelectorActionMiddleware should throw error for invalid payload type 1`] = `"Selector actions should be based on action creator of type function, instead received the following type: [undefined]"`;
